### ISSUE TITLE
Impressions from my experience getting started with B-Tax

### ISSUE THREAD

**Issue by martinholmer** [open-source-economics/B-Tax#218]
I followed a variant of the "**first way** to use B-Tax" that did not download source code but rather installed the btax conda package.  I did this as an experiment to see if the package I made for the B-Tax 0.2.1 release works the same way as the package for B-Tax 0.2.2, which was not made by me.  So, when reading this comment remember that I have no B-Tax source code on my computer and that the experiment was conducted in a working directory that has no access to the Tax-Calculator source code that is on my computer.

My experiment involved executing the `run_btax_example.py` script, which is what the B-Tax README file suggests as the way to get started with B-Tax.  Doing this required me to copy the `2017_law.json` and `run_btax_example.py` files into an empty working directory and also copy to that directory the `puf.csv` input data file.  Then I installed the btax package for release 0.2.2 (which is the latest version) this way:
```
iMac:btax mrh$ conda install -c OSPC btax=0.2.2 --yes
Solving environment: done
## Package Plan ##
  environment location: /Users/mrh/anaconda3
  added / updated specs: 
    - btax=0.2.2
The following NEW packages will be INSTALLED:
    btax:    0.2.2-py36_0  OSPC
    taxcalc: 0.22.2-py36_0 OSPC
Preparing transaction: done
Verifying transaction: done
Executing transaction: done
```
Next I executed the `run_btax_example.py` script and got this:
```
iMac:btax mrh$ python run_btax_example.py 
You loaded data for 2014.
Tax-Calculator startup automatically extrapolated your data to 2014.
/Users/mrh/anaconda3/lib/python3.6/site-packages/pandas/util/_decorators.py:178: FutureWarning: the 'skip_footer' keyword is deprecated, use 'skipfooter' instead
  return func(*args, **kwargs)
/Users/mrh/anaconda3/lib/python3.6/site-packages/pandas/core/frame.py:6211: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version
of pandas will change to not sort by default.

To accept the future behavior, pass 'sort=False'.

To retain the current behavior and silence the warning, pass 'sort=True'.

  sort=sort)
/Users/mrh/anaconda3/lib/python3.6/site-packages/pandas/io/excel.py:329: FutureWarning: The `sheetname` keyword is deprecated, use `sheet_name` instead
  **kwds)
Dump asset_data.pkl
Loaded asset_data.pkl
You loaded data for 2011.
Tax-Calculator startup automatically extrapolated your data to 2013.
{'tau_nc': 0.25007793488098795, 'tau_div': 0.18284089327179814, 'tau_int': 0.30819693255690533, 'tau_scg': 0.3347690108009861, 'tau_lcg': 0.21418606313653052, 'tau_td': 0.21744734204085464, 'tau_h': 0.019922892489679392}
/Users/mrh/anaconda3/lib/python3.6/site-packages/pandas/core/frame.py:6211: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version
of pandas will change to not sort by default.

To accept the future behavior, pass 'sort=False'.

To retain the current behavior and silence the warning, pass 'sort=True'.

  sort=sort)
Loaded asset_data.pkl
You loaded data for 2011.
Tax-Calculator startup automatically extrapolated your data to 2013.
{'tau_nc': 0.3159348001807159, 'tau_div': 0.1921063832327825, 'tau_int': 0.33127250075114056, 'tau_scg': 0.35739009652293235, 'tau_lcg': 0.22013627233133193, 'tau_td': 0.2506651871607095, 'tau_h': 0.14856444983426845}
```
More on the warnings later in this comment.

And here is the output I got:
```
iMac:btax mrh$ ls -l
  drwxr-xr-x  13 mrh       442 Nov  6 07:53 .
  drwxr-xr-x  51 mrh      1734 Nov  6 07:46 ..
  -rw-r--r--   1 mrh      2651 Nov  6 07:49 2017_law.json <-- INSTALLED BY ME
  -rw-r--r--   1 mrh  19908136 Nov  6 07:51 asset_data.pkl <-- WHAT IS THIS?
  -rw-r--r--   1 mrh     66926 Nov  6 07:53 base_byasset.csv
  -rw-r--r--   1 mrh     55575 Nov  6 07:53 baseline_byindustry.csv <-- MISNAMED
  drwxr-xr-x   2 mrh        68 Nov  6 07:51 btax_output_dir <-- EMPTY; WHY IS THIS HERE?
  -rw-r--r--   1 mrh     73281 Nov  6 07:53 changed_byasset.csv
  -rw-r--r--   1 mrh     53718 Nov  6 07:53 changed_byindustry.csv
  -rwxr-xr-x   1 mrh  56415698 Nov  6 07:51 puf.csv <-- INSTALLED BY ME
  -rw-r--r--   1 mrh     73729 Nov  6 07:53 reform_byasset.csv
  -rw-r--r--   1 mrh     55359 Nov  6 07:53 reform_byindustry.csv
  -rw-r--r--   1 mrh      1042 Nov  6 07:50 run_btax_example.py <-- INSTALLED BY ME
```
In the above annotated directory listing, I've marked the three files I installed and raise some questions about the output.  The B-Tax README lists only six output files.  So, I don't understand what the `asset_data.pkl` file contains.  Also, the README says one of the output files should be `base_byindustry.csv`, but I got `baseline_byindustry.csv`; seems like it is an easy fix to drop the "line" from that output file's name.  And my final question is why is the `btax_output_dir` created when nothing is written to that directory?  Seems like that should be eliminated or the six output files should be written to that subdirectory; one or the other.

I did the above procedure installing the btax package for B-Tax release 0.2.2 (as described above) and then, after cleaning up the 0.2.2 results, I did the above procedure installing the btax package for B-Tax release 0.2.1, which I created a couple of days ago.  It turns out the the format of the output files changed with the addition of seven new variables in the `byasset` files and maybe changes in the `byindustry` files (but I didn't look at those in detail).  When I compared the subset of `byasset` output variables that were in both the 0.2.1 and 0.2.2 versions, I found almost complete agreement.  But there were a few differences: namely in the values of the `z_c` and `z_c_d` variables.  Here's the first set of of differences (with the 0.2.1 results on the left and the 0.2.2 results on the right):
<img width="1392" alt="screen shot 2018-11-06 at 11 10 13 am" src="https://user-images.githubusercontent.com/12170745/48077363-2de34780-e1b5-11e8-9847-022b0d21c5b9.png">

I'm don't know nearly enough about B-Tax to have an informed view on whether these non-trivial differences are caused by improvements in data and/or code in version 0.2.2 relative to version 0.2.1.

The fact that the vast majority of the `byasset` results are exactly the same when using the two different packages indicates, I think, that the new package-building procedure that I used to build the B-Tax 0.2.1 package is working as designed.  So, that's the end of the package-building experiment.

Let me conclude with a few comments on the warnings that were generated by B-Tax when I executed `run_btax_example.py`.

**First**, the famous pandas sort warning:
```
/Users/mrh/anaconda3/lib/python3.6/site-packages/pandas/core/frame.py:6211:
FutureWarning: Sorting because non-concatenation axis is not aligned.
A future version of pandas will change to not sort by default.
To accept the future behavior, pass 'sort=False'.
To retain the current behavior and silence the warning, pass 'sort=True'.
```
@andersonfrailey and I had to deal with this one months ago in the taxdata and Tax-Calculator code.
I've completely forgotten how we handled this warning, which comes from a `pd.concat()` call as I remember, but I'm sure @andersonfrailey can give you some advice about how to handle this matter.  You can also search the Internet; there's been plenty of discussion of this warning.

**Second**, a pandas warning that `skipfooter` should be used (rather than `skip_footer`):
```
/Users/mrh/anaconda3/lib/python3.6/site-packages/pandas/util/_decorators.py:178:
FutureWarning: the 'skip_footer' keyword is deprecated, use 'skipfooter' instead
```
I have no idea why this comes up in B-Tax, but it seems easy to fix.

**Third**, yet another pandas warning when doing Excel spreadsheet IO:
```
/Users/mrh/anaconda3/lib/python3.6/site-packages/pandas/io/excel.py:329:
FutureWarning: The `sheetname` keyword is deprecated, use `sheet_name` instead
```
Again, I have no idea why B-Tax is doing Excel input/output, but the fix seems easy enough.

These second and third pandas warnings are par for the course and are why some get exasperated with pandas developers.  **Why** are they moving in opposite directions?  They want you make these changes:
```
skip_footer ---> skipfooter 
sheetname   ---> sheet_name
```
Where's the consistency?  At least its easy to do a find and replace in your code.

One final thought.  It would be ideal to have the six output files expected to be generated by the `run_btax_example.py` script in the repository (in the `run_examples` or should that be `run_example` subdirectory).  That way new users will have confidence that they can get the same results on their computers as the B-Tax developers get on their computers.

**Comment by jdebacker** [comment]
@martinholmer Thank you for taking the time to build the package and give this thorough account of what you found.  It looks like you had a successful build.

Let me respond to a few of your remarks:

> Also, the README says one of the output files should be base_byindustry.csv, but I got baseline_byindustry.csv; seems like it is an easy fix to drop the "line" from that output file's name

Indeed.  Thank you for catching this.

> And my final question is why is the btax_output_dir created when nothing is written to that directory? Seems like that should be eliminated or the six output files should be written to that subdirectory; one or the other.

That's right.  This used to be used, but the `mkdir` was never taken out of the codebase.  This should be removed.

> I'm don't know nearly enough about B-Tax to have an informed view on whether these non-trivial differences are caused by improvements in data and/or code in version 0.2.2 relative to version 0.2.1.

These diffs are a result of changes in assumptions regarding how intellectual property can be depreciated under the tax code that resulted from reading recent CBO work on the topic.  They are unrelated to your build.

> Let me conclude with a few comments on the warnings that were generated by B-Tax

I think most of these are now fixed with PR #215.  They had not occurred to me until recently as I had been using a stale environment (something we discussed in another issue over in the OG-USA repo).

> One final thought. It would be ideal to have the six output files expected to be generated by the run_btax_example.py script in the repository (in the run_examples or should that be run_example subdirectory). That way new users will have confidence that they can get the same results on their computers as the B-Tax developers get on their computers.

Ok - this makes sense.  Would you suggest adding something to the `run_btax_example.py` script to show the differences between the files checked into the repo and those created by the user?

**Comment by martinholmer** [comment]
@jdebacker asked in B-Tax issue #218:
>> One final thought. It would be ideal to have the six output files expected to be generated by the run_btax_example.py script in the repository (in the run_examples or should that be run_example subdirectory). That way new users will have confidence that they can get the same results on their computers as the B-Tax developers get on their computers.
>
> Ok - this makes sense. Would you suggest adding something to the `run_btax_example.py` script to show the differences between the files checked into the repo and those created by the user?

Probably not needed.  What is needed are different names for the "expected" results files.
So, for example, the `reform_byasset.csv` generated by the `run_btax_example.py` script will overwrite the "expected" results file if the "expected" results file has the same name.  To avoid this, name the "expected" results file something like `reform_byasset_expected.csv`.  Then users can simply use their favorite diff utility to see if there are differences between the "actual" `reform_byasset.csv` file contents and the "expected" `reform_byasset_expected.csv` file contents.

**Comment by jdebacker** [comment]
@martinholmer thanks for your input. I'll follow your advice here.

**Comment by jdebacker** [comment]
The specific issues discussed here to be fixed were noted in Issue #219 and resolved with PR #220.
